### PR TITLE
Code Review by Nicola

### PR DESCRIPTION
**Pull request by bangndnguyen**

### CODE REVIEW

File: client/src/app/compose/compose.component.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            selected: false
+        };
+        this.to = new FormControl(this.mail.to, [
+            Validators.required
+        ]);
+        this.title = new FormControl(this.mail.title, [
+            Validators.required
+        ]);
+        this.message = new FormControl(this.mail.message);
+        this.resetForm();
```
Comment (nicola-recruit) [inline client/src/app/compose/compose.component.ts]: If you use the form builder of reactive forms, you can reduce this boiler plate code a lot; the form object that the reactive form builder returns provides a nice `reset` method that you use after submittal.

File: client/src/app/compose/compose.component.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    public send (): void {
+        this.db.collection('mails').add({
+            from: this.auth.me.email,
```
Comment (nicola-recruit) [inline client/src/app/compose/compose.component.ts]: Here two things are going on: the creation of a new JSON element plus the processing of its insertion.
Maybe could be nice to split the creation in one function and keep the store operation here.

File: client/src/app/core/auth.service.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    public signIn (email: string, password: string): void {
+        this.afAuth.auth.signInWithEmailAndPassword(email, password);
+    }
+
+    public signOut (): void {
+        this.afAuth.auth.signOut();
+    }
+
+    public signUp (email: string, password: string, username: string): void {
```
Comment (nicola-recruit) [inline client/src/app/core/auth.service.ts]: Authentication and Sign-up are two complete different operations that are not necessarily related to each other.
I suggest to split responsibilities and move this method into a new, dedicated service.

File: client/src/app/core/auth.service.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        this.afAuth.auth.createUserWithEmailAndPassword(email, password).then((userCredential) => {
+            this.afs.doc('users/' + userCredential.user.uid).set({
+                uid: userCredential.user.uid,
+                role: 'member',
+                email: userCredential.user.email,
+                username: username
+            });
```
Comment (nicola-recruit) [inline client/src/app/core/auth.service.ts]: Maybe is better to create the object in a specific method and let this `signup` method to do only the signup operation.

File: client/src/app/core/auth.service.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    constructor (private afAuth: AngularFireAuth, private afs: AngularFirestore) {
+        this.meObservable = this.afAuth.authState.pipe(switchMap((fbUser) => {
+            if (fbUser) {
+                return this.afs.doc<User>('users/' + fbUser.uid).valueChanges();
+            }
+            return [null];
+        }));
+        this.meObservable.subscribe((user: User) => {
+            this.me = user;
+        });
```
Comment (nicola-recruit) [inline client/src/app/core/auth.service.ts]: I am ignorant about Firebase, so this is the first time I read code for using its API. I must say it is not easy at all, so I can understand why you had a hard time with it.
However, I think that whenever we choose a library, it is good practice to encapsulate it into one or more services that expose an API that we could use even if we change the backing library.
In this case, I think there should be at least one service that encapsulate Firebase API and exposes simpler and more readable methods for the rest of the application services.
In other words, don't use directly Firebase directly in your service but put it behind one more service (thin layer).

File: client/src/app/inbox/inbox.component.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+                        id: changeAction.payload.doc.id,
+                        ...changeAction.payload.doc.data()
+                    };
+                });
+            })) : empty();
+        })).subscribe((mails) => {
+            this.mails = mails;
+            this.currentMail = mails[0];
+        });
+    }
```
Comment (nicola-recruit) [inline client/src/app/inbox/inbox.component.ts]: I have no idea what's going on here... too many moving parts, no simpler functions, naming hard to understand.

File: client/src/app/inbox/inbox.component.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+    public onDeleteButtonClick(): void {
+        const selectedMails = this.mails.filter(mail => mail.selected);
+        if (selectedMails.length > 0) {
+            const dialogRef = this.dialog.open(DeleteMailDialogComponent, {
+                width: '250px',
+                data: {},
+                panelClass: 'panel-class',
+            });
+            dialogRef.afterClosed().subscribe(result => {
```
Comment (nicola-recruit) [inline client/src/app/inbox/inbox.component.ts]: Two different things happening in one function. Good idea could be to split them out into different methods.

File: client/src/app/navbar/navbar.component.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+@Component({
+    selector: 'app-navbar',
+    templateUrl: './navbar.component.html',
+    styleUrls: ['./navbar.component.css']
+})
+export class NavbarComponent implements OnInit {
+    constructor(private router: Router, private auth: AuthService) {}
+
+    public ngOnInit(): void {
+        this.auth.meObservable.subscribe((user) => {
```
Comment (nicola-recruit) [inline client/src/app/navbar/navbar.component.ts]: `meObservable` stands for... Sorry but the naming is not clear enough.

File: client/src/app/navbar/navbar.component.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+            this.redirect(user);
+        });
+    }
+
+    private redirect (user: User): void {
+        if (user) {
+            this.router.navigate(['/inbox']);
+        } else {
+            this.router.navigate(['/']);
```
Comment (nicola-recruit) [inline client/src/app/navbar/navbar.component.ts]: I would encapsulate the router into a specific service and use its API to define where I want to go.
If we change URL structure, you just touch one point and you're done.

File: client/src/app/signup/signup.component.ts [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    styleUrls: ['./signup.component.scss']
+})
+export class SignupComponent implements OnInit {
+    public signupForm: FormGroup;
+
+    constructor (private auth: AuthService, private formBuilder: FormBuilder, private afs: AngularFirestore) {}
+
+    public ngOnInit(): void {
+        this.signupForm = this.formBuilder.group({
+            email: [
```
Comment (nicola-recruit) [inline client/src/app/signup/signup.component.ts]: Ah! You are using the form builder here, nice!

File: client/src/app/signup/signup.component.ts [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+    get username () {
+        return this.signupForm.get('username');
+    }
+
+    get email () {
+        return this.signupForm.get('email');
+    }
+
+    get password () {
```
Comment (nicola-recruit) [inline client/src/app/signup/signup.component.ts]: Interesting usage of `get`!

File: client/src/app/signup/signup.component.ts [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+        return this.signupForm.get('password');
+    }
+}
+
+class CustomValidators {
```
Comment (nicola-recruit) [inline client/src/app/signup/signup.component.ts]: Why two classes in one file? You could import how many you want seamlessly...

File: client/src/app/test.ts [hunk 13]
```diff
@@ -0,0 +1,2 @@
+export class Test {
```
Comment (nicola-recruit) [inline client/src/app/test.ts]: ? Empty?
